### PR TITLE
add config options to support loading arbitrary scripts (with permission support)

### PR DESCRIPTION
**Pull request by twrichards**
Co-authored-by: @tjsilver 

https://trello.com/c/TIeyuEff/511-1st-pinboard-integration-grid

## What does this change?
- New config option in `KahunaConfig`(i.e. `kahuna.conf`) entitled `scriptsToLoad` (list of objects) which allows the loading of arbitrary scripts (immediate use case is https://github.com/guardian/editorial-tools-pinboard) which optionally have a permission requirement on them, for example...
image

- Refactors to make the `Content-Security-Policy` more configurable by...
   - adding new config option `security.connectSources` (list of string) to control the `connect-src` portion of the CSP
   - re-using the 'hosts' from the new `scriptsToLoad` config option in the `script-src` portion of the CSP

## Tested?
- [ ] locally by committer
- [ ] locally by Guardian reviewer
- [ ] on the Guardian's TEST environment

### CODE REVIEW

**Comment by sihil** [comment]
Ah - totally understand our conversation the other day now. Remind me again, why is the pinboard client not being rolled into [gustaf](https://github.com/guardian/gustaf)?

**Comment by twrichards** [comment]
> Ah - totally understand our conversation the other day now. Remind me again, why is the pinboard client not being rolled into [gustaf](https://github.com/guardian/gustaf)?

Because we didn't want to have to wait for changes to the extension to propagate through to people's browsers, when we release pinboard (currently around 3sec deploy ⚡️) it's instantly available for all users. We could consider moving the loader part to gustaf but it works neatly as is and the integration is pretty light touch and neatly all behind a permission.

Also worth reminding that there is no teleporter for grid (appreciate I haven't checked if it in fact still runs on grid and just doesn't render the Ⓣ icon because there's no unique ID).

**Comment by twrichards** [comment]
@sihil do you think this change is relatively open-source friendly (i.e. we're not coupling pinboard and grid here, just making it possible to add it if desired)?

**Comment by sihil** [comment]
It is loosely coupled and the open source element is well dealt with.

My concern stems from the subtle increase in the number of update sources that could break the grid. It is becomes feasible to make and deploy a change to pinboard that makes grid unusable. As a pattern it concerns me for the same reason that I dislike dependency definitions that are not pinned to a specific version: I'm spent too much time hunting down issues caused by a subtle change to a dependency. In short, there is a cost to making pinboard quick to deploy which is increasing the brittleness of the grid as a system - it's not as self contained.

Gustaf already falls into this category but is arguably a well known exception (presence does too, but in a rebuild I'd change it to be a normal library). For gustaf, it's easy to prove/disprove it being a problem by disabling the extension, although I guess we can block the pinboard script for the same debugging experience.

Whilst the integration is light touch, it's also light touch across all of the systems you want pinboard on. Gustaf (which does have [permission](https://github.com/guardian/gustaf/blob/master/extension/manifest.json#L18) could be even easier, although I don't think you can dynamically load JS into an extension so rollout would indeed be slower.

**Comment by twrichards** [comment]
@sihil 

> It is loosely coupled and the open source element is well dealt with.

👍

> My concern stems from the subtle increase in the number of update sources that could break the grid. It is becomes feasible to make and deploy a change to pinboard that makes grid unusable. As a pattern it concerns me for the same reason that I dislike dependency definitions that are not pinned to a specific version: I'm spent too much time hunting down issues caused by a subtle change to a dependency. In short, there is a cost to making pinboard quick to deploy which is increasing the brittleness of the grid as a system - it's not as self contained.
> 
> Gustaf already falls into this category but is arguably a well known exception (presence does too, but in a rebuild I'd change it to be a normal library). For gustaf, it's easy to prove/disprove it being a problem by disabling the extension, although I guess we can block the pinboard script for the same debugging experience.

I would argue that if a problem is reported its less hassle for user for us to redeploy a safe version of Pinboard (sub 5 seconds) than have the user mess about disabling extensions, and in the process loosing Teleporter functionality, then waiting for a revert release (or fix-forward release) of gustaf to roll out.

> Whilst the integration is light touch, it's also light touch across all of the systems you want pinboard on. Gustaf (which does have [permission](https://github.com/guardian/gustaf/blob/master/extension/manifest.json#L18) could be even easier, although I don't think you can dynamically load JS into an extension so rollout would indeed be slower.

Pinboard itself already has permission integration as the main line of defence, but the permission stuff in this PR (and in https://github.com/guardian/flexible-content/pull/3706) is simply to avoid loading JS that isn't going to be allowed to run (and in the process keeps load on pinboard's `bootstrapping-lambda` to a minimum). I appreciate gustaf has permission built in already but pinboards current approach allows us to iterate quickly.

File: rest-lib/src/main/scala/com/gu/mediaservice/lib/auth/provider/AuthorisationProvider.scala
Comment (sihil) [inline rest-lib/src/main/scala/com/gu/mediaservice/lib/auth/provider/AuthorisationProvider.scala:6]: The Guardian permissions library has leaked into the abstraction here. This more or less defeats the point so I'm afraid I won't allow this. I suggest that as an alternative we create a temporary permission, perhaps `PinboardAssetsFeature` with an appropriate comment on it and an intention to get rid of it at some point in the future.

I plan to refactor the codebase in the coming months so that this sort of leak isn't possible (Guardian specific code lives in a separate sbt subproject so these common libraries can't see the dependency).

Comment (twrichards) [reply rest-lib/src/main/scala/com/gu/mediaservice/lib/auth/provider/AuthorisationProvider.scala:6]: my subconscious was telling me this might be problematic. happy to do that, as it reduces the complexity. Do you think the config model introduced in this PR should (for the time being) remove the notion of `app` so the `permission` field associated with a script in `scriptsToLoad` should just be a String (which will be checked against the `name` field of the `SimplePermission`?

Comment (sihil) [reply rest-lib/src/main/scala/com/gu/mediaservice/lib/auth/provider/AuthorisationProvider.scala:6]: Or even hardcoded to just work with pinboard for now. I do think we should pursue a proper feature switch system to solve all of these problems, although it might be that when we do bring permissions with params back in that might also work for this.

**Review by jonathonherbert — APPROVED** [review]
Code looks fine, LGTM and other commenters with more informed opinions re: specific parts have expressed approval out of band – 👍

**Pull request merged**
